api/v1/testclustertopology_webhook.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var testclustertopologylog = logf.Log.WithName("testclustertopology-resource")
+
+func (r *TestClusterTopology) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+// +kubebuilder:webhook:path=/mutate-naglfar-pingcap-com-v1-testclustertopology,mutating=true,failurePolicy=fail,groups=naglfar.pingcap.com,resources=testclustertopologies,verbs=create;update,versions=v1,name=mtestclustertopology.kb.io
+
+var _ webhook.Defaulter = &TestClusterTopology{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *TestClusterTopology) Default() {
+	testclustertopologylog.Info("default", "name", r.Name)
+
+	// TODO(user): fill in your defaulting logic.
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+// +kubebuilder:webhook:verbs=create;update,path=/validate-naglfar-pingcap-com-v1-testclustertopology,mutating=false,failurePolicy=fail,groups=naglfar.pingcap.com,resources=testclustertopologies,versions=v1,name=vtestclustertopology.kb.io
+
+var _ webhook.Validator = &TestClusterTopology{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *TestClusterTopology) ValidateCreate() error {
+	testclustertopologylog.Info("validate create", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object creation.
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *TestClusterTopology) ValidateUpdate(old runtime.Object) error {
+	testclustertopologylog.Info("validate update", "name", r.Name)
+	topology := old.(*TestClusterTopology)
+
+	if topology.Status.State == "updating" {
+		return errors.New("cluster is updating")
```

Comment:
How to resolve the situation when you set an invalid configuration so the tiup fails to reload? You should give the user a path to revise the incorrect configration.